### ISSUE TITLE
AOT support for C# directors

### ISSUE THREAD

**Issue by KallynGowdyYetiCGI** [swig/swig#1165]
Hello, I'm currently trying to use SWIG as a tool to generate C++ bindings for C# that is used in Unity 3D for mobile development. Of particular note is that I need to support calling C# code from C++.

Enter directors.

Unfortunately, the code that SWIG generates does not work in scenarios where JIT is not available because of the use of instance member methods attached to delegates. 

To explain it another way, when run on iOS (which disallows JIT compilation) IL2CPP throws an exception because it detects that I'm trying to marshal a function pointer from unmanaged code. (explanation below)

In normal .Net, the VM would simply emit the code necessary to handle the callback at runtime, but with IL2CPP (the backend that Unity 3D uses) we need to tell them about the callback ahead of time so that they can generate the code at compile time. We can do this with `MonoPInvokeCallbackAttribute`, but this has the noticeable drawback of only supporting callbacks to managed methods that are static. (See [this page](https://docs.unity3d.com/Manual/TroubleShootingIPhone.html) for more info; search for `"MonoPInvoke"`) 

This is where my major problem comes in. Declaring a class in C# that overrides methods declared by a class in C++ simply cannot work without changing how SWIG generates the C++ and C# director code.

Effectively, I need SWIG to generate directors that accept an additional pointer when calling `swig_connect_director(delegate)`. This pointer would be used as a way to reference methods that are tied to object instances.

For example, `swig_connect_director` would be redefined as:
```c++
void swig_connect_director(SWIG_Callback callback, void *delegate_ref) {
  swig_callback = callback;
  swig_delegate_ref = delegate_ref;
}
```
where `swig_delegate_ref` is a `void *` field on the `SwigDirector_XyzBase` class and `SWIG_Callback` is defined as:
```c++
typedef void (SWIGSTDCALL* SWIG_Callback0_t)(void *, args...);
```
(pointer to a function whose arguments accept `void *` and whatever arguments the actual method requires)

Then, on the C# side we would generate a static function which wraps the instance delegate.
```c#
// New delegate to represent the static method callback.
public delegate void SwigDelegateStaticAction (IntPtr delegate_ref, args...);

// Existing delegate which SWIG already creates
public delegate void SwigDelegateAction(args...);

// Instead of using SwigDirectorAction as the callback,
// we use a static method which will resolve the instance method for us.
[MonoPInvokeCallback]
private static void SwigDirectorStaticAction(IntPtr delegate_ref, args...)
{
  var handle = GCHandle.FromIntPtr(delegate_ref);
  var action = (SwigDelegateAction)handle.Target;
  action(args...);
}

// Existing function
private void SwigDirectorAction(args...) 
{
  Action(args...);
}

// Reworked to support passing the delegate handle to C++
private void SwigDirectorConnect()
{
  swigDelegate = new SwigDelegateAction(SwigDirectorUpdate);
  var handle = GCHandle.Alloc(swigDelegate);
  var delegate_ref = GCHandle.ToIntPtr(handle);

  // Notice that we pass the static method reference because IL2CPP
  // requires that the callback is static.
  PINVOKE.MyClass_director_connect(swigCPtr, SwigDirectorStaticAction, delegate_ref);
}

private SwigDelegateAction swigDelegate;
```
See [this blog post](http://blog.reblochon.org/2013/03/working-around-reverse-callback.html) for the inspiration.

**So, long story short**, this is something that I need to get fixed and would be able to develop myself (assuming I don't find an easier/better solution). I just need to know 2 things:

1. Is there a way to override this functionality without forking SWIG myself? If so, what features would I be looking at and are there any examples that I could learn from?
2. If I do need to fork it, where would you recommend starting? I've already looked at the `monotouch` branch and, while it starts the work towards what I want, I fear it is too out of date for me to pick up where the author left off.

**Comment by wsfulton** [comment]
I suggest you aim to create a patch for merging into mainline SWIG and proceed as follows:

1. Create your own branch off latest master.
2. Merge in the monotouch branch.
3. Add your changes and test.
4. Create a pull request for us to review for future inclusion in SWIG. 

The monotouch branch stalled because no-one verified it worked. Have a look at the emails at https://sourceforge.net/p/swig/mailman/message/35847193/. When you've added your suggested fixes, can you also suggest a way to test it on either Windows, Linux or MacOSX? We can then set up Travis to test it so that your changes don't regress.

With regard to your proposed fixes, can you use HandleRef instead of IntPtr in the unmanaged function declarations as this fixes early garbage collection issues. Well, this applies when calling from managed code to unmanaged, but I'm not entirely sure when going the other way, but better to be safe than sorry.

**Comment by ojwb** [comment]
Closing as a duplicate of #1262.
